In [1]:
import numpy as np
import pandas as pd
import requests

In [2]:
#read in text file

with open('/Users/anandigupta/Desktop/coding_discussions_ppol564_fall2021/04_coding_discussion/Data/aljazeera-khashoggi.txt') as f:
    aljazeera = f.read()
aljazeera

'Turkey, Istanbul - Turkish President Recep Tayyip Erdogan has said the murder of journalist Jamal Khashoggi at the kingdom\'s consulate in Istanbul was planned by Saudi officials days in advance. Addressing legislators from his Justice and Development Party (AK Party) on Tuesday, Erdogan detailed Khashoggi\'s disappearance and murder but stopped short of accusing Saudi royals of the "savage" killing that has caused global outrage. "On September 28, Khashoggi arrived at the Saudi Arabian consulate for him to sort out his wedding paperwork," Erdogan said during the speech in the Turkish parliament in the capital, Ankara. "It seems that at that time they [Saudi Arabian officials] started to plan a roadmap for his murder." He added that some Saudi officials left Turkey and travelled to Saudi Arabia, "indicating they planned the murder". Khashoggi, 59, a Washington Post columnist and critic of the powerful Saudi Crown Prince Mohammed bin Salman, disappeared after entering the Saudi consula

In [3]:
#read in text file

with open('/Users/anandigupta/Desktop/coding_discussions_ppol564_fall2021/04_coding_discussion/Data/bbc-khashoggi.txt') as f:
    bbc = f.read()
bbc

'Recep Tayyip Erdogan says many questions need to be answered The killing of journalist Jamal Khashoggi was planned days in advance, Turkish President Recep Tayyip Erdogan has told MPs from his ruling party. He said Turkey had strong evidence Khashoggi was killed in a premeditated and "savage" murder at the Saudi consulate in Istanbul on 2 October. He also called for the suspects to be tried in Istanbul. He demanded Saudi Arabia provide answers about where Khashoggi\'s body was, and who had ordered the operation. The Saudi kingdom has provided conflicting accounts of what happened to Khashoggi, a US resident and Washington Post contributor. After weeks of maintaining he was still alive, the authorities now say the 59-year-old was killed in a rogue operation. Mr Erdogan\'s address coincided with the start of an investment conference in Saudi Arabia that has been overshadowed by the Khashoggi case. Dozens of government and business leaders have pulled out, but Crown Prince Mohammed bin S

In [4]:
#read in text file

with open('/Users/anandigupta/Desktop/coding_discussions_ppol564_fall2021/04_coding_discussion/Data/breitbart-khashoggi.txt') as f:
    breitbart = f.read()
breitbart

'U.S. President Donald Trump and Turkish President Recep Tayyip Erdogan both pronounced themselves unsatisfied with the answers Saudi Arabia has given for the killing of journalist Jamal Khashoggi at the Saudi embassy in Istanbul three weeks ago. Both said the Saudis must be more forthcoming about who ordered the operation that ended with Khashoggi’s death, followed by crude efforts to create an alibi and dispose of his body. Erdogan was forceful in a speech to his AKP party on Tuesday, although he came up short of accusing Saudi King Salman bin Abdulaziz or Crown Prince Mohammed bin Salman of direct involvement with the killing. “It is clear that this savage murder did not happen instantly but was planned,” Erdogan declared, dismissing the Saudi story about Khashoggi accidentally dying in a “fistfight” he started when he visited the Istanbul consulate. “Why was a team of 15 Saudi men, all with qualifications related to the incident, gathering in Istanbul on the day of the murder?” the

In [5]:
#read in text file

with open('/Users/anandigupta/Desktop/coding_discussions_ppol564_fall2021/04_coding_discussion/Data/cnn-khashoggi.txt') as f:
    cnn = f.read()
cnn


'Istanbul, Turkey (CNN) Jamal Khashoggi died as a result of a brutal premeditated murder, Turkey\'s President Recep Tayyip Erdogan said on Tuesday, in a highly anticipated speech in which he soundly rejected Saudi Arabia\'s claim that the journalist was killed accidentally.Erdogan questioned whether Riyadh was able to carry out an independent investigation in to the death of the Washington Post journalist, and he demanded that the 18 suspects arrested by authorities in Saudi Arabia be extradited to Turkey to stand trial in Istanbul.It was the first time that any official in Turkey has publicly outlined the Turkish contention that Khashoggi was killed at Saudi Arabia\'s consulate in Istanbul by a hit squad sent from Riyadh. While Erdogan did not offer any bombshell revelations, his speech amounted to a comprehensive rejection of Saudi Arabia\'s case that Khashoggi died by accident as a result of a brawl.People gather around a TV at an Istanbul cafe to watch Erdogan\'s speech."The inform

In [6]:
#read in text file

with open('/Users/anandigupta/Desktop/coding_discussions_ppol564_fall2021/04_coding_discussion/Data/fox-khashoggi.txt') as f:
    fox = f.read()
fox

'Saudi officials planned the savage murder of Saudi writer Jamal Khashoggi days before his death, Turkish President Recep Tayyip Erdogan said Tuesday. Erdogan revealed the details of the country’s investigation into Khashoggi’s killing after he walked into the Saudi consulate in Istanbul on Oct. 2, contradicting Saudi Arabia’s explanation that the writer was killed in a “fistfight.” Erdogan fell short of blaming Saudi Crown Prince Mohammed bin Salman and made no mention of whether a tape exists of Khashoggi\'s killing. However, he kept the pressure on the kingdom with his demands for punishment of all the people involved in the murder. “All evidence gathered shows that Jamal Khashoggi was the victim of a savage murder. To cover up such a savagery would hurt the human conscience,” he said. Erdogan said he wanted Saudi Arabia to allow the 18 suspects detained in the columnist’s killing to be tried in Turkish courts. “To blame such an incident on a handful of security and intelligence mem

In [7]:
#basic cleaning and splitting of text

def tokenize(text=None):
    text = text.lower()
    text = text.replace('.','')
    text_list = text.split()
    return text_list

In [8]:
def convert_text_to_dtm(txt):
    '''
    Converts text into a document term matrix.
    '''
    d = dict()
    for word in tokenize(txt):
        if word in d:
            d[word][0] += 1
        else:
            d[word] = [1]
    return pd.DataFrame(d)

convert_text_to_dtm(cnn)

istanbul,  turkey  (cnn)  jamal  khashoggi  died  as   a  result  of  ...  \
0          2       3      1      1          5     2   2  14       2   6  ...   

   were  scouted,  but  noted  later  khashoggi's  body  yet  foundread  more  
0     1         1    1      1      1            1     1    1          1     1  

[1 rows x 151 columns]

In [9]:
# Now build a function that does this for a list of texts
def gen_DTM(texts=None):
    '''
    Generate a document term matrix
    '''
    DTM = pd.DataFrame()
    for text in texts:
        entry = convert_text_to_dtm(text)
        DTM = DTM.append(pd.DataFrame(entry),ignore_index=True,sort=True) # Row bind
    
    DTM.fillna(0, inplace=True) # Fill in any missing values with 0s (i.e. when a word is in one text but not another)
    return DTM

#apply to the five news articles     
gen_DTM([aljazeera, bbc, breitbart, fox, cnn]) 

"a  "all  "assassination  "be  "bring  "co-operator"  "distinction"  \
0  1.0   0.0             1.0  0.0     0.0            0.0            1.0   
1  0.0   1.0             0.0  1.0     1.0            1.0            0.0   
2  0.0   0.0             0.0  0.0     0.0            0.0            0.0   
3  0.0   1.0             0.0  0.0     0.0            0.0            0.0   
4  0.0   0.0             0.0  0.0     0.0            0.0            0.0   

   "everything  "fist-fight"  "fistfight"  ...  “may  “other  “partner  \
0          1.0           0.0          1.0  ...   0.0     0.0       0.0   
1          0.0           1.0          0.0  ...   0.0     0.0       0.0   
2          0.0           0.0          0.0  ...   1.0     1.0       1.0   
3          0.0           0.0          0.0  ...   0.0     0.0       0.0   
4          0.0           0.0          0.0  ...   0.0     0.0       0.0   

   “putting  “saudi  “the  “to  “we  “we're  “why  
0       0.0     0.0   0.0  0.0  0.0     0.0   0.0  
1       0.0     0.0   0.0  0.0  0.0     0.0   0.0  
2       1.0     0.0   2.0  0.0  1.0     0.0   1.0  
3       0.0     1.0   0.0  1.0  0.0     1.0   0.0  
4       0.0     0.0   0.0  0.0  0.0     0.0   0.0  

[5 rows x 1016 columns]

In [10]:
#calculate cosines using numpy

def cosine(a,b):
    cos = np.dot(a,b)/(np.sqrt(np.dot(a,a)) * np.sqrt(np.dot(b,b))  )
    return cos

In [11]:
#generate document term matrix for all five articles and save to dictionary

def calc_corr():
    

    D = gen_DTM([aljazeera, bbc, breitbart, fox,cnn]) 
    data = {'aljazeera': D.iloc[0].values,
        'bbc': D.iloc[1].values,
        'breitbart': D.iloc[2].values,
        'fox': D.iloc[3].values,
        'cnn': D.iloc[4].values
        }

    df = pd.DataFrame(data,columns=['aljazeera','bbc','breitbart','fox','cnn'])

#calculate cosines for all combinations of news sources

    corrMatrix = df.corr(method = cosine)
    return (corrMatrix)
    

In [12]:
calc_corr()

aljazeera       bbc  breitbart       fox       cnn
aljazeera   1.000000  0.862252   0.824850  0.832189  0.730518
bbc         0.862252  1.000000   0.891241  0.886869  0.745456
breitbart   0.824850  0.891241   1.000000  0.868990  0.683575
fox         0.832189  0.886869   0.868990  1.000000  0.743866
cnn         0.730518  0.745456   0.683575  0.743866  1.000000

In [13]:
#read in list of stop words

stop_words = pd.read_csv('/Users/anandigupta/Desktop/coding_discussions_ppol564_fall2021/04_coding_discussion/Data/stop_words.csv')

#convert to list 

stop_words = stop_words.word.to_list()
stop_words

['a',
 "a's",
 'able',
 'about',
 'above',
 'according',
 'accordingly',
 'across',
 'actually',
 'after',
 'afterwards',
 'again',
 'against',
 "ain't",
 'all',
 'allow',
 'allows',
 'almost',
 'alone',
 'along',
 'already',
 'also',
 'although',
 'always',
 'am',
 'among',
 'amongst',
 'an',
 'and',
 'another',
 'any',
 'anybody',
 'anyhow',
 'anyone',
 'anything',
 'anyway',
 'anyways',
 'anywhere',
 'apart',
 'appear',
 'appreciate',
 'appropriate',
 'are',
 "aren't",
 'around',
 'as',
 'aside',
 'ask',
 'asking',
 'associated',
 'at',
 'available',
 'away',
 'awfully',
 'b',
 'be',
 'became',
 'because',
 'become',
 'becomes',
 'becoming',
 'been',
 'before',
 'beforehand',
 'behind',
 'being',
 'believe',
 'below',
 'beside',
 'besides',
 'best',
 'better',
 'between',
 'beyond',
 'both',
 'brief',
 'but',
 'by',
 'c',
 "c'mon",
 "c's",
 'came',
 'can',
 "can't",
 'cannot',
 'cant',
 'cause',
 'causes',
 'certain',
 'certainly',
 'changes',
 'clearly',
 'co',
 'com',
 'come',
 'c

In [14]:
#redefine cleaning function to remove words in stop word list

def tokenize(text=None):
    text = text.lower()
    text = text.replace('.','')
    text_list = text.split()
    text_list2 = [word for word in text_list if word not in stop_words]
    return text_list2

In [15]:
#recalculate cosines after removing stop words

calc_corr()

aljazeera       bbc  breitbart       fox       cnn
aljazeera   1.000000  0.636264   0.525267  0.603883  0.500149
bbc         0.636264  1.000000   0.533582  0.574735  0.470131
breitbart   0.525267  0.533582   1.000000  0.495539  0.325097
fox         0.603883  0.574735   0.495539  1.000000  0.490545
cnn         0.500149  0.470131   0.325097  0.490545  1.000000

In [16]:
#redefine cleaning function to only keep alphabets

def tokenize(text=None):
    text = text.lower()
    text = text.replace('.','')
    text_list = text.split()
    text_list2 = [''.join(e for e in string if e.isalpha()) for string in text_list]
    text_list3 = [word for word in text_list2 if word not in stop_words]
    return text_list3

In [17]:
#recalculate cosines after removing numbers and special characters

calc_corr()


aljazeera       bbc  breitbart       fox       cnn
aljazeera   1.000000  0.686191   0.597133  0.712376  0.533815
bbc         0.686191  1.000000   0.597764  0.644908  0.505989
breitbart   0.597133  0.597764   1.000000  0.530848  0.380198
fox         0.712376  0.644908   0.530848  1.000000  0.520635
cnn         0.533815  0.505989   0.380198  0.520635  1.000000

In [18]:
#find max and min cosines
c = calc_corr().abs()

s = c.unstack()
so = s.sort_values(kind="quicksort")
so

breitbart  cnn          0.380198
cnn        breitbart    0.380198
           bbc          0.505989
bbc        cnn          0.505989
cnn        fox          0.520635
fox        cnn          0.520635
           breitbart    0.530848
breitbart  fox          0.530848
cnn        aljazeera    0.533815
aljazeera  cnn          0.533815
breitbart  aljazeera    0.597133
aljazeera  breitbart    0.597133
breitbart  bbc          0.597764
bbc        breitbart    0.597764
fox        bbc          0.644908
bbc        fox          0.644908
           aljazeera    0.686191
aljazeera  bbc          0.686191
           fox          0.712376
fox        aljazeera    0.712376
aljazeera  aljazeera    1.000000
breitbart  breitbart    1.000000
bbc        bbc          1.000000
fox        fox          1.000000
cnn        cnn          1.000000
dtype: float64

In general, there seems to be a fair amount of similarity between how the Khashoggi scandal was reported across the five different sources. Aljazeera and Fox seem to be the most similar (cosine of 0.71 after removing stop words, numbers, and special characters) and Breitbart and CNN seem to be the least similar (cosine of 0.38 after removing stop words, numbers, and special characters). 

Cosines are highest with minimal cleaning (before removing stop words, numbers, or special characters). This is because all of the news sources likely use words that are common to most sentences, and this is inflating how similar the sources appear. Removing stop words lower the cosines substantially. However, once we remove numbers and other special characters, we see cosines increase slightly. This also makes intuitive sense, as these characters are generally meaningless, but differences in how numbers and characters are used by different authors will make the sources appear less similar to each other than they truly are.